File: src/pes_exploration/idwm.rs
//! Interatomic distance weighted metric
use crate::pes_exploration::system::System;
use crate::pes_exploration::traits::PES;
use ndarray::{Array1, Array2, array, Axis, s};





fn w(x: f64) -> f64
{
    let n: i32 = 5;
    let sigma: f64 = 3.0;

    ( -(x/sigma).powi(n) ).exp() + 1.0
}

fn dw(x: f64) -> f64
{
    let n: i32 = 5;
    let sigma: f64 = 3.0;

    ( -(x/sigma).powi(n) ).exp() * ( -(n as f64) * x.powi(n-1) / sigma.powi(n) )
}





/// Input the atomic coordinates of a system (coord),
/// output the weighted distance matrix.
///
/// # Parameters
/// ```
/// coord: the atomic coordinates of the system (natom * 3)
/// w_mat: the weighted distance matrix (natom * natom)
/// ```
pub fn wei_dist_mat(coord: &Array2<f64>) -> Array2<f64>
{
    let natom: usize = coord.len_of(Axis(0));

    let mut w_mat: Array2<f64> = Array2::zeros((natom, natom));
    for i in 0..(natom-1)
    {
        for j in (i+1)..natom
        {
            w_mat[[i,j]] = w( ( (coord[[i,0]]-coord[[j,0]]).powi(2) + (coord[[i,1]]-coord[[j,1]]).powi(2) + (coord[[i,2]]-coord[[j,2]]).powi(2) ).sqrt() );
        }
    }

    return w_mat
}

/// Input the weighted distance matrix of a reference system (w_mat),
/// and the atomic coordinates of another system (coord),
/// output the interatomic distance weighted (IDW) metric between them.
///
/// # Parameters
/// ```
/// w_mat0: the weighted distance matrix of the reference system (natom * natom)
/// coord: the atomic coordinates of the system (natom * 3)
/// dist: the output IDW distance between the two systems
/// ```
pub fn idw_dist(w_mat0: &Array2<f64>, coord: &Array2<f64>) -> f64
{
    let natom: usize = coord.len_of(Axis(0));

    let mut dist: f64 = 0.0;
    for i in 0..(natom-1)
    {
        for j in (i+1)..natom
        {
            dist += (w( ( (coord[[i,0]]-coord[[j,0]]).powi(2) + (coord[[i,1]]-coord[[j,1]]).powi(2) + (coord[[i,2]]-coord[[j,2]]).powi(2) ).sqrt() ) - w_mat0[[i,j]]).powi(2);
        }
    }

    return dist.sqrt()
}

pub fn idw_dist1(w_mat0: &Array2<f64>, w_mat1: &Array2<f64>) -> f64
{
    let natom: usize = w_mat0.len_of(Axis(0));

    let mut dist: f64 = 0.0;
    for i in 0..(natom-1)
    {
        for j in (i+1)..natom
        {
            dist += (w_mat1[[i,j]] - w_mat0[[i,j]]).powi(2);
        }
    }

    return dist.sqrt()
}

pub fn idw_dist_vec(w_mat0: &Array2<f64>, coord: &Array2<f64>) -> (f64, Array2<f64>)
{
    let natom: usize = coord.len_of(Axis(0));

    let mut vec_mat: Array2< Array1<f64> > = Array2::from_elem((natom, natom), array![0.0, 0.0, 0.0]);
    let mut dist_mat: Array2<f64> = Array2::zeros((natom, natom));
    let mut w_mat: Array2<f64> = Array2::zeros((natom, natom));
    let mut dw_mat: Array2<f64> = Array2::zeros((natom, natom));
    let mut dist: f64 = 0.0;
    for i in 0..(natom-1)
    {
        for j in (i+1)..natom
        {
            vec_mat[[i,j]] = &coord.slice(s![i, ..]) - &coord.slice(s![j, ..]);            // the (i, j) element is a vector from atom j to atom i
            dist_mat[[i,j]] = (vec_mat[[i,j]][0].powi(2) + vec_mat[[i,j]][1].powi(2) + vec_mat[[i,j]][2].powi(2)).sqrt();           // the (i, j) element is the atomic distance between atom i and j
            w_mat[[i,j]] = w(dist_mat[[i,j]]);
            dw_mat[[i,j]] = dw(dist_mat[[i,j]]);
            dist += (w_mat[[i,j]] - w_mat0[[i,j]]).powi(2);
        }
    }
    dist = dist.sqrt();

    let mut vec: Array2<f64> = Array2::zeros((natom, 3));
    for i in 0..natom
    {
        for j in 0..natom
        {
            if i<j
            {
                vec[[i,0]] += (w_mat[[i,j]] - w_mat0[[i,j]]) * dw_mat[[i,j]] * vec_mat[[i,j]][0] / dist_mat[[i,j]];
                vec[[i,1]] += (w_mat[[i,j]] - w_mat0[[i,j]]) * dw_mat[[i,j]] * vec_mat[[i,j]][1] / dist_mat[[i,j]];
                vec[[i,2]] += (w_mat[[i,j]] - w_mat0[[i,j]]) * dw_mat[[i,j]] * vec_mat[[i,j]][2] / dist_mat[[i,j]];
            }
            if j<i
            {
                vec[[i,0]] -= (w_mat[[j,i]] - w_mat0[[j,i]]) * dw_mat[[j,i]] * vec_mat[[j,i]][0] / dist_mat[[j,i]];
                vec[[i,1]] -= (w_mat[[j,i]] - w_mat0[[j,i]]) * dw_mat[[j,i]] * vec_mat[[j,i]][1] / dist_mat[[j,i]];
                vec[[i,2]] -= (w_mat[[j,i]] - w_mat0[[j,i]]) * dw_mat[[j,i]] * vec_mat[[j,i]][2] / dist_mat[[j,i]];
            }
        }
    }
    vec /= dist;

    return (dist, vec)
}





fn idw_pot(w_mat0: &Array2<f64>, coord: &Array2<f64>, a: f64, sigma: f64) -> f64
{
    let idw_dist: f64 = idw_dist(w_mat0, coord);
    let pot_idw: f64 = a * ( -(idw_dist*idw_dist) / (2.0*sigma*sigma) ).exp();

    return pot_idw
}

fn idw_pot_force(w_mat0: &Array2<f64>, coord: &Array2<f64>, a: f64, sigma: f64) -> (f64, Array2<f64>)
{
    let (idw_dist, idw_vec) = idw_dist_vec(w_mat0, coord);
    let pot_idw: f64 = a * ( -(idw_dist*idw_dist) / (2.0*sigma*sigma) ).exp();
    let force_idw: Array2<f64> = idw_vec * pot_idw * idw_dist / (sigma*sigma);

    return (pot_idw, force_idw)
}





pub struct Idwm0PES
{
    pub local_min: Array2<f64>,
    pub nearby_ts: Vec<Array2<f64>>,
    pub a_min: f64,
    pub a_ts: f64,
    pub sigma_min: f64,
    pub sigma_ts: Vec<f64>,
}

impl PES for Idwm0PES
{
    fn get_energy(&self, s: &System) -> f64
    {
        match &s.atom_add_pot
        {
            // If the input structure do not specify the atoms to be added IDWM potential
            None =>
            {
                let mut pot_idwm: f64 = idw_pot(&self.local_min, &s.coord, self.a_min, self.sigma_min);
                for i in 0..self.nearby_ts.len()
                {
                    pot_idwm += idw_pot(&self.nearby_ts[i], &s.coord, self.a_ts, self.sigma_ts[i]);
                }

                return pot_idwm
            },

            // If the input structure specifies the atoms to be added IDWM potential
            Some(atom_add_pot) =>
            {
                let mut s_fragment_coord: Array2<f64> = Array2::zeros((atom_add_pot.len(), 3));
                for i in 0..atom_add_pot.len()
                {
                    s_fragment_coord[[i, 0]] = s.coord[[atom_add_pot[i], 0]];
                    s_fragment_coord[[i, 1]] = s.coord[[atom_add_pot[i], 1]];
                    s_fragment_coord[[i, 2]] = s.coord[[atom_add_pot[i], 2]];
                }

                let mut pot_idwm: f64 = idw_pot(&self.local_min, &s_fragment_coord, self.a_min, self.sigma_min);
                for i in 0..self.nearby_ts.len()
                {
                    pot_idwm += idw_pot(&self.nearby_ts[i], &s_fragment_coord, self.a_ts, self.sigma_ts[i]);
                }

                return pot_idwm
            },
        }
    }

    fn get_energy_force(&self, s: &System) -> (f64, Array2<f64>)
    {
        match &s.atom_add_pot
        {
            // If the input structure do not specify the atoms to be added IDWM potential
            None =>
            {
                let (mut pot_idwm, mut force_idwm) = idw_pot_force(&self.local_min, &s.coord, self.a_min, self.sigma_min);
                let mut pot_idwm_ts: f64;
                let mut force_idwm_ts: Array2<f64>;
                for i in 0..self.nearby_ts.len()
                {
                    (pot_idwm_ts, force_idwm_ts) = idw_pot_force(&self.nearby_ts[i], &s.coord, self.a_ts, self.sigma_ts[i]);
                    pot_idwm += pot_idwm_ts;
                    force_idwm += &force_idwm_ts;
                }

                return (pot_idwm, force_idwm)
            },

            // If the input structure specifies the atoms to be added IDWM potential
            Some(atom_add_pot) =>
            {
                let mut s_fragment_coord: Array2<f64> = Array2::zeros((atom_add_pot.len(), 3));
                for i in 0..atom_add_pot.len()
                {
                    s_fragment_coord[[i, 0]] = s.coord[[atom_add_pot[i], 0]];
                    s_fragment_coord[[i, 1]] = s.coord[[atom_add_pot[i], 1]];
                    s_fragment_coord[[i, 2]] = s.coord[[atom_add_pot[i], 2]];
                }

                let (mut pot_idwm, mut force_idwm_fragment) = idw_pot_force(&self.local_min, &s_fragment_coord, self.a_min, self.sigma_min);
                let mut pot_idwm_ts: f64;
                let mut force_idwm_ts: Array2<f64>;
                for i in 0..self.nearby_ts.len()
                {
                    (pot_idwm_ts, force_idwm_ts) = idw_pot_force(&self.nearby_ts[i], &s_fragment_coord, self.a_ts, self.sigma_ts[i]);
                    pot_idwm += pot_idwm_ts;
                    force_idwm_fragment += &force_idwm_ts;
                }

                let mut force_idwm: Array2<f64> = Array2::zeros(s.coord.raw_dim());
                for i in 0..atom_add_pot.len()
                {
                    force_idwm[[atom_add_pot[i], 0]] = force_idwm_fragment[[i, 0]];
                    force_idwm[[atom_add_pot[i], 1]] = force_idwm_fragment[[i, 1]];
                    force_idwm[[atom_add_pot[i], 2]] = force_idwm_fragment[[i, 2]];
                }

                return (pot_idwm, force_idwm)
            },
        }
    }
}

File: rtipmd/jax/src/rtip_jax/core/idwm.py
"""Interatomic distance weighted metric.

Rust source: `src/pes_exploration/idwm.rs`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Sequence

import jax.numpy as jnp

from rtip_jax.system import System

IDWM_WEIGHT_N = 5
IDWM_WEIGHT_SIGMA = 3.0


def weight(distance: Any) -> Any:
    """Rust `w(x) = exp(-(x / 3)^5) + 1`."""

    distance = jnp.asarray(distance, dtype=jnp.float64)
    return jnp.exp(-((distance / IDWM_WEIGHT_SIGMA) ** IDWM_WEIGHT_N)) + 1.0


def weight_derivative(distance: Any) -> Any:
    """Derivative of `weight` with respect to distance."""

    distance = jnp.asarray(distance, dtype=jnp.float64)
    return jnp.exp(-((distance / IDWM_WEIGHT_SIGMA) ** IDWM_WEIGHT_N)) * (
        -float(IDWM_WEIGHT_N) * (distance ** (IDWM_WEIGHT_N - 1)) / (IDWM_WEIGHT_SIGMA**IDWM_WEIGHT_N)
    )


def _as_coord(coord: Any) -> Any:
    coord = jnp.asarray(coord, dtype=jnp.float64)
    if coord.ndim != 2 or coord.shape[1] != 3:
        raise ValueError(f"coord must have shape (natom, 3), got {coord.shape}")
    return coord


def _as_square_matrix(matrix: Any, natom: int | None = None) -> Any:
    matrix = jnp.asarray(matrix, dtype=jnp.float64)
    if matrix.ndim != 2 or matrix.shape[0] != matrix.shape[1]:
        raise ValueError(f"matrix must have shape (natom, natom), got {matrix.shape}")
    if natom is not None and matrix.shape[0] != natom:
        raise ValueError(f"matrix shape {matrix.shape} does not match natom={natom}")
    return matrix


def _upper_mask(natom: int) -> Any:
    return jnp.triu(jnp.ones((natom, natom), dtype=bool), k=1)


def _pairwise_vectors(coord: Any) -> Any:
    return coord[:, None, :] - coord[None, :, :]


def _pairwise_distances(coord: Any) -> Any:
    return jnp.linalg.norm(_pairwise_vectors(coord), axis=-1)


def wei_dist_mat(coord: Any) -> Any:
    """Return the upper-triangular weighted distance matrix.

    The lower triangle and diagonal are zero, matching the Rust implementation.
    """

    coord = _as_coord(coord)
    natom = coord.shape[0]
    mask = _upper_mask(natom)
    return jnp.where(mask, weight(_pairwise_distances(coord)), 0.0)


def idw_dist(w_mat0: Any, coord: Any) -> Any:
    """Return IDW distance between a reference weighted matrix and coordinates."""

    coord = _as_coord(coord)
    w_mat0 = _as_square_matrix(w_mat0, coord.shape[0])
    mask = _upper_mask(coord.shape[0])
    current = wei_dist_mat(coord)
    return jnp.sqrt(jnp.sum(jnp.where(mask, (current - w_mat0) ** 2, 0.0)))


def idw_dist1(w_mat0: Any, w_mat1: Any) -> Any:
    """Return IDW distance between two weighted distance matrices."""

    w_mat0 = _as_square_matrix(w_mat0)
    w_mat1 = _as_square_matrix(w_mat1, w_mat0.shape[0])
    mask = _upper_mask(w_mat0.shape[0])
    return jnp.sqrt(jnp.sum(jnp.where(mask, (w_mat1 - w_mat0) ** 2, 0.0)))


def idw_dist_vec(w_mat0: Any, coord: Any) -> tuple[Any, Any]:
    """Return IDW distance and the corresponding coordinate vector.

    This preserves Rust's singular behavior: if the IDW distance is zero, the
    final normalization produces NaN values instead of applying an epsilon.
    """

    coord = _as_coord(coord)
    w_mat0 = _as_square_matrix(w_mat0, coord.shape[0])
    natom = coord.shape[0]
    mask = _upper_mask(natom)

    vec_mat = _pairwise_vectors(coord)
    dist_mat = _pairwise_distances(coord)
    w_mat = jnp.where(mask, weight(dist_mat), 0.0)
    dw_mat = jnp.where(mask, weight_derivative(dist_mat), 0.0)
    delta = jnp.where(mask, w_mat - w_mat0, 0.0)
    dist = jnp.sqrt(jnp.sum(delta**2))

    safe_dist_mat = jnp.where(mask, dist_mat, 1.0)
    coeff = jnp.where(mask, delta * dw_mat / safe_dist_mat, 0.0)
    pair_vec = coeff[:, :, None] * vec_mat
    vec = pair_vec.sum(axis=1) - pair_vec.sum(axis=0)
    vec = vec / dist

    return dist, vec


def idw_pot(w_mat0: Any, coord: Any, a: float, sigma: float) -> Any:
    """Return IDWM Gaussian potential energy."""

    dist = idw_dist(w_mat0, coord)
    return float(a) * jnp.exp(-(dist * dist) / (2.0 * float(sigma) * float(sigma)))


def idw_pot_force(w_mat0: Any, coord: Any, a: float, sigma: float) -> tuple[Any, Any]:
    """Return IDWM Gaussian potential energy and force."""

    dist, vec = idw_dist_vec(w_mat0, coord)
    pot = float(a) * jnp.exp(-(dist * dist) / (2.0 * float(sigma) * float(sigma)))
    force = vec * pot * dist / (float(sigma) * float(sigma))
    return pot, force


@dataclass(frozen=True)
class Idwm0PES:
    """IDWM repulsive potential-energy surface.

    `local_min` and `nearby_ts` are weighted distance matrices, not coordinates,
    matching Rust `Idwm0PES`.
    """

    local_min: Any
    nearby_ts: Sequence[Any]
    a_min: float
    a_ts: float
    sigma_min: float
    sigma_ts: Sequence[float]

    def __post_init__(self) -> None:
        local_min = _as_square_matrix(self.local_min)
        nearby_ts = tuple(_as_square_matrix(item, local_min.shape[0]) for item in self.nearby_ts)
        sigma_ts = tuple(float(value) for value in self.sigma_ts)
        if len(nearby_ts) != len(sigma_ts):
            raise ValueError("nearby_ts and sigma_ts must have the same length")
        object.__setattr__(self, "local_min", local_min)
        object.__setattr__(self, "nearby_ts", nearby_ts)
        object.__setattr__(self, "a_min", float(self.a_min))
        object.__setattr__(self, "a_ts", float(self.a_ts))
        object.__setattr__(self, "sigma_min", float(self.sigma_min))
        object.__setattr__(self, "sigma_ts", sigma_ts)

    def _active_coord(self, system: System) -> tuple[Any, tuple[int, ...] | None]:
        if system.atom_add_pot is None:
            return system.coord, None
        indices = tuple(system.atom_add_pot)
        return system.coord[jnp.asarray(indices)], indices

    def get_energy(self, system: System) -> float:
        coord, _indices = self._active_coord(system)
        energy = idw_pot(self.local_min, coord, self.a_min, self.sigma_min)
        for ts_matrix, sigma in zip(self.nearby_ts, self.sigma_ts, strict=True):
            energy = energy + idw_pot(ts_matrix, coord, self.a_ts, sigma)
        return float(energy)

    def get_energy_force(self, system: System) -> tuple[float, Any]:
        coord, indices = self._active_coord(system)
        energy, force_fragment = idw_pot_force(self.local_min, coord, self.a_min, self.sigma_min)
        for ts_matrix, sigma in zip(self.nearby_ts, self.sigma_ts, strict=True):
            ts_energy, ts_force = idw_pot_force(ts_matrix, coord, self.a_ts, sigma)
            energy = energy + ts_energy
            force_fragment = force_fragment + ts_force

        if indices is None:
            return float(energy), force_fragment

        force = jnp.zeros_like(system.coord)
        force = force.at[jnp.asarray(indices)].set(force_fragment)
        return float(energy), force


__all__ = [
    "IDWM_WEIGHT_N",
    "IDWM_WEIGHT_SIGMA",
    "Idwm0PES",
    "idw_dist",
    "idw_dist1",
    "idw_dist_vec",
    "idw_pot",
    "idw_pot_force",
    "wei_dist_mat",
    "weight",
    "weight_derivative",
]

